### PR TITLE
[plex] Initial contribution

### PR DESCRIPTION
**Pull request by aronbeurskens**
# New binding: Plex

# Description
This adds a new binding for communication with Plex.
It allows multiple players to be added to OpenHab and provide information about what is currently playing on the play.

When using this binding one can:
- Create pages with meta information about the media that is begin played
- Create rules to automatically control lights when a movie is started / paused / stopped

# Related links
https://github.com/openhab/openhab-addons/issues/6179
https://github.com/openhab/openhab-addons/issues/4949
https://community.openhab.org/t/plex-binding-for-oh3-will-it-be-ported/111487

# Testing
Latest .jar file available add: https://github.com/aronbeurskens/openhab-addons/releases/tag/v0.13

### CODE REVIEW

**Comment by aronbeurskens** [comment]
Can somebody help/point me solving the build failures? It seems to fail, but not on the plex part.
How can I reproduce this locally (what command to run) to debug what is causing the failed build?

**Comment by lolodomo** [comment]
Maybe due to the full copyright you added in comments in the file feature.cml? 
Look at another binding how should look like this file.

**Comment by aronbeurskens** [comment]
> 

Thanks, I will have a look. Is there any way to reproduce this locally, to avoid having to commit and wait each time...

**Comment by lolodomo** [comment]
Yes, by running mvn clean install at the root of your git trpo.

**Comment by aronbeurskens** [comment]
Thank you @lolodomo . Managed to get it passing the build checks!

**Comment by aronbeurskens** [comment]
This is my first contribution / pr so I'm not quite familiar with the way of working. Do I need to do anything to get this merged?

**Comment by fwolter** [comment]
Seems like there went something wrong when you force-pushed your changes. There are files changed outside of your binding. 

See https://community.openhab.org/t/rebase-your-code-or-how-to-fix-your-git-history-before-requesting-a-pull/129358 for fixing it.

**Review by jlaur — CHANGES_REQUESTED** [review]
Updating review state to reflect reality.

**Comment by SeveranExp** [comment]
What’s the status of this binding? Seems to be stuck for months. Can we get this moving somehow?

**Comment by lolodomo** [comment]
> What’s the status of this binding? Seems to be stuck for months. Can we get this moving somehow?

The PR is currently invalid as it proposes changes in many bindings.

**Comment by SeveranExp** [comment]
> > What’s the status of this binding? Seems to be stuck for months. Can we get this moving somehow?
> 
> The PR is currently invalid as it proposes changes in many bindings.

Understood, but letting it stuck like this is not helping any. Is there anything we can do to unstuck this?

**Comment by fwolter** [comment]
Sure, the failed rebase needs to be cleaned up. See https://community.openhab.org/t/rebase-your-code-or-how-to-fix-your-git-history-before-requesting-a-pull/129358 how to do this. If you want to do that, you first need to fork the contributor's branch and apply the fix on your forked branch. Then, you can file a new PR based on your branch.

**Comment by aronbeurskens** [comment]
I'll rebase my branch and update the PR.

**Comment by aronbeurskens** [comment]
Code = rebased
Checks passed

Merge is blocked by change request, do i need to do anything to get it unblocked?

**Comment by SeveranExp** [comment]
Seems to me that the changes requested by @jlaur are triggering review requests by 12 other people… @fwolter can you provide explanation on how to untangle this?

**Comment by fwolter** [comment]
The PR is now ready for review. There's nothing which needs to be done by you.

**Comment by jlaur** [comment]
> Seems to me that the changes requested by @jlaur are triggering review requests by 12 other people

No, this was triggered by the faulty merge including code from other bindings. These reviewers were then added as code owners because of the included changes. They can be removed manually.

File: bundles/org.openhab.binding.plex/README.md
Comment (mlobstein) [inline bundles/org.openhab.binding.plex/README.md:3]: This is the placeholder README and must be updated for the binding. The README is the main source of documentation for the binding and will be displayed on the add-ons page on the OH site.

Comment (SeveranExp) [reply bundles/org.openhab.binding.plex/README.md:3]: @mlobstein who updates the read me?

**Comment by aronbeurskens** [comment]
Im willing to complete it. I think I can find some time next week

**Comment by SeveranExp** [comment]
> 

If you need a review let me know and I’ll assist.

File: bundles/org.openhab.binding.plex/README.md
Comment (mlobstein) [reply bundles/org.openhab.binding.plex/README.md:3]: The binding author is responsible for providing the readme.

**Comment by aronbeurskens** [comment]
Documentation is updated. @SeveranExp please review it to get this show on the road.

**Comment by SeveranExp** [comment]
> Documentation is updated. @SeveranExp please review it to get this show on the road.

Read me reviewed, nothing to add from my perspective. We can always add more examples as time goes on and community comes up with more examples.
Let’s do this!

**Comment by aronbeurskens** [comment]
Hi all,

You've asked me several weeks ago to put in some effort to fix this PR which I did. Can you put in some effort to get this PR merged?

**Comment by SeveranExp** [comment]
@hmerk sorry to bug you, but is there something that is missing from our end?? How long should we expect to have this merged??

**Comment by hmerk** [comment]
@SeveranExp As I am not a maintainer, I cannot tell.
You have to wait until our maintainers find the time to fully review this PR.
This is not the only one in the queue….

**Comment by SeveranExp** [comment]
> @SeveranExp As I am not a maintainer, I cannot tell.
> You have to wait until our maintainers find the time to fully review this PR.
> This is not the only one in the queue….
> In the meantime, it can be added to the community marketplace, so users can easily install and test it.

That’s interesting. How can this be done??

**Comment by hmerk** [comment]
@SeveranExp see here
https://community.openhab.org/t/announcing-the-community-marketplace/127188

File: bundles/org.openhab.binding.plex/pom.xml
Comment (wborn) [inline bundles/org.openhab.binding.plex/pom.xml:10]:
```suggestion
    <version>3.4.0-SNAPSHOT</version>
```

File: bundles/org.openhab.binding.plex/README.md
Comment (fwolter) [inline bundles/org.openhab.binding.plex/README.md:55]: MainUI specific documentation should be avoided. You could simply say, that the auto discovery can be started.

File: bundles/org.openhab.binding.plex/README.md
Comment (fwolter) [inline bundles/org.openhab.binding.plex/README.md:101]:
```suggestion
| currentPlayers       | String | RO         | The number of players currently configured to watch on PLEX     |
```

File: bundles/org.openhab.binding.plex/README.md
Comment (fwolter) [inline bundles/org.openhab.binding.plex/README.md:102]: These seem rather `Number` than `String`.

File: bundles/org.openhab.binding.plex/README.md
Comment (fwolter) [inline bundles/org.openhab.binding.plex/README.md:103]: You could list the possible values of this channel.

File: bundles/org.openhab.binding.plex/README.md
Comment (fwolter) [inline bundles/org.openhab.binding.plex/README.md:104]: Can this be `Switch` instead of `String`?

File: bundles/org.openhab.binding.plex/README.md
Comment (fwolter) [inline bundles/org.openhab.binding.plex/README.md:107]: This could be rather `DateTime`.

File: bundles/org.openhab.binding.plex/README.md
Comment (fwolter) [inline bundles/org.openhab.binding.plex/README.md:111]: This could be of type `Player`. See https://www.openhab.org/docs/configuration/items.html#type

File: bundles/org.openhab.binding.plex/README.md
Comment (fwolter) [inline bundles/org.openhab.binding.plex/README.md:172]: The readme should contain any specific UI details. You could post example use cases in the samples section of the community forum. See https://community.openhab.org/c/tutorials-examples/10

**Review by fwolter — CHANGES_REQUESTED** [review]
I reviewed the readme, so far.

File: bundles/org.openhab.binding.plex/README.md
Comment (aronbeurskens) [reply bundles/org.openhab.binding.plex/README.md:102]: It is used in:
 
```
public ChannelUID(     @NotNull org.openhab.core.thing.ThingUID thingUID,
    String id )
```

which needs a String as argument.

File: bundles/org.openhab.binding.plex/README.md
Comment (aronbeurskens) [reply bundles/org.openhab.binding.plex/README.md:104]: See above: It is the name of the channel itself, not the actual value. So it should be a string.

File: bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexApiConnector.java
Comment (fwolter) [inline bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexApiConnector.java:171]: When an error occured during connecting, you could set the ThingStatus to OFFLINE with the ThingStatusDetail to COMMUNICATION_ERROR or CONFIGURATION_ERROR. The error message you are currently logging here can then be set as the thing status detail message. This will be displayed in the UI, showing the user what's wrong. See `updateStatus()` in the ThingHandler.

File: bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexApiConnector.java
Comment (fwolter) [inline bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexApiConnector.java:290]: Can this be removed?

File: bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexApiConnector.java
Comment (fwolter) [inline bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexApiConnector.java:401]: Bindings should only log to error if something severe happened, like the detection of a bug in your code. This could be warn. See this link for a description of the log levels: https://www.openhab.org/docs/developer/guidelines.html#f-logging

File: bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexPlayerHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexPlayerHandler.java:76]: This is actually handled by the framework. You need to set the thing status to UNKNOWN in this method. Then, the Thing is automatically set ONLINE/OFFLINE according the bridge's status.

File: bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexServerHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexServerHandler.java:99]: The log message could be removed, as the status update is already logged by the framework.

The state change originated by updateStatus() is logged to events.log. Including the status detail message.

Please check all.

File: bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexServerHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexServerHandler.java:135]: Can this be removed?

File: bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexServerHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexServerHandler.java:56]: This is never used and can therefore be removed.

File: bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexServerHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexServerHandler.java:268]: The field can be null at this point, because it isn't inittialized in the constructor. So, you need to check if it is not null before accessing it: Example: if (job != null). But this would result in a compiler warning, as it is a field and could theoretically be modified by another thread exactly between the null check and the access in the next line. To mitigate this, you can store the field to a local variable before checking it:
```
Future<?> localPollingJob = pollingJob;
if (localPollingJob != null) {
    localPollingJob.cancel(true);
}
```

Please check all.

File: bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexServerHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexServerHandler.java:292]: The exception's message could be used as the thing status detail message. Then, the log message could be removed and the error is displayed in the UI.

File: bundles/org.openhab.binding.plex/src/main/resources/OH-INF/binding/binding.xml
Comment (fwolter) [inline bundles/org.openhab.binding.plex/src/main/resources/OH-INF/binding/binding.xml:8]: The author tag is deprecated and should therefore be removed. See openhab/openhab-core#1844.

File: bundles/org.openhab.binding.plex/src/main/resources/OH-INF/config/config.xml
Comment (fwolter) [inline bundles/org.openhab.binding.plex/src/main/resources/OH-INF/config/config.xml:48]:
```suggestion
			<description>The unique identifier of the player</description>
```

File: bundles/org.openhab.binding.plex/src/main/resources/OH-INF/i18n/plex.properties
Comment (fwolter) [inline bundles/org.openhab.binding.plex/src/main/resources/OH-INF/i18n/plex.properties:2]: Can these comments be removed?

You could create this default translation file by running `mvn org.openhab.core.tools:i18n-maven-plugin:3.3.0:generate-default-translations` in your binding's directory. Then, [Crowdin](https://crowdin.com/project/openhab-addons) will pick up the English "translation" and users can help to translate your binding in Crowdin. If you want you can use Crowdin, too to translate your binding as soon as this PR is merged.

File: bundles/org.openhab.binding.plex/src/main/resources/OH-INF/i18n/plex.properties
Comment (fwolter) [inline bundles/org.openhab.binding.plex/src/main/resources/OH-INF/i18n/plex.properties:15]: Can these be removed?

File: bundles/org.openhab.binding.plex/src/main/resources/OH-INF/thing/thing-types.xml
Comment (fwolter) [inline bundles/org.openhab.binding.plex/src/main/resources/OH-INF/thing/thing-types.xml:7]: Can this be removed?

**Review by fwolter — CHANGES_REQUESTED** [review]
There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html` after building it with `mvn clean install`.

There are some compiler warnings about null access left. You could take a look at `mvn clean install`. To fix these, you need to store the field to a local variable and do the null check on that.

Are the screenshots under `doc/` used somewhere anymore?

**Comment by SeveranExp** [comment]
@aronbeurskens hey man! Noticed you’ve been away for a while. Hope you’re doing okay. Tagging you to ask if you could find some time from your day to review the changes proposed above?
Cheers!

**Comment by aronbeurskens** [comment]
Hi @SeveranExp , thanks for tagging me, i somehow missed the feedback / status. I'll make some time in the next weeks to process the feedback.

**Comment by SeveranExp** [comment]
@aronbeurskens no worries :) excited about the possibility of having this available soon :)

**Comment by SeveranExp** [comment]
No update? :( @aronbeurskens

**Comment by aronbeurskens** [comment]
Hi all, it has been a while, but letting you all know i am working on an update compatible for version 4.0

File: bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/discovery/PlexDiscoveryService.java
Comment (aronbeurskens) [reply bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/discovery/PlexDiscoveryService.java:37]: At this point I am failing to see how this would make things easier. But probably due to my lack of knowledge at this point. 

The bridge in this case only shows the things when they are active (for instance playing a movie). This is done by the method `getAvailablePlayers`.

If this should be refactored, could you help me in the right direction?

File: bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexServerHandler.java
Comment (aronbeurskens) [reply bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexServerHandler.java:270]: I don't see this case. If the first condition is met `pollingJob == null` it doesn't execute the next `pollingJob.isCancelled()`. In the next lines the `pollingJob` is overwritten.

**Comment by aronbeurskens** [comment]
@fwolter thanks for the feedback. I've updated the PR as much as possible. Please have second look at it.

File: bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexServerHandler.java
Comment (morph166955) [inline bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexServerHandler.java:292]: I saw this while compiling a jar of this to play with.  Is System.out.printf() appropriate here?  I would think it should be new String() or something similar to that.

**Review by morph166955 — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.plex/src/main/resources/OH-INF/binding/binding.xml
Comment (morph166955) [reply bundles/org.openhab.binding.plex/src/main/resources/OH-INF/binding/binding.xml:1]: I built a jar to test this on 4.0.0 and had to resolve this to do so.  I can't commit to the branch but this diff should fix this issue.

```
diff --git a/bundles/org.openhab.binding.plex/src/main/resources/OH-INF/addon/addon.xml b/bundles/org.openhab.binding.plex/src/main/resources/OH-INF/addon/addon.xml
new file mode 100644
index 0000000000..ddc096213f
--- /dev/null
+++ b/bundles/org.openhab.binding.plex/src/main/resources/OH-INF/addon/addon.xml
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<addon:addon id="plex" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xmlns:addon="https://openhab.org/schemas/addon/v1.0.0"
+       xsi:schemaLocation="https://openhab.org/schemas/addon/v1.0.0 https://openhab.org/schemas/addon-1.0.0.xsd">
+
+       <type>binding</type>
+       <name>Plex Binding</name>
+       <description>Monitor your plex server and players</description>
+
+</addon:addon>
diff --git a/bundles/org.openhab.binding.plex/src/main/resources/OH-INF/binding/binding.xml b/bundles/org.openhab.binding.plex/src/main/resources/OH-INF/binding/binding.xml
deleted file mode 100644
index a690db4c86..0000000000
--- a/bundles/org.openhab.binding.plex/src/main/resources/OH-INF/binding/binding.xml
+++ /dev/null
@@ -1,9 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
-<binding:binding id="plex" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-       xmlns:binding="https://openhab.org/schemas/binding/v1.0.0"
-       xsi:schemaLocation="https://openhab.org/schemas/binding/v1.0.0 https://openhab.org/schemas/binding-1.0.0.xsd">
-
-       <name>Plex Binding</name>
-       <description>Monitor your plex server and players</description>
-
-</binding:binding>

```

File: bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexServerHandler.java
Comment (aronbeurskens) [reply bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexServerHandler.java]: Thanks for the feedback, I've updated the code

File: bundles/org.openhab.binding.plex/src/main/resources/OH-INF/binding/binding.xml
Comment (aronbeurskens) [reply bundles/org.openhab.binding.plex/src/main/resources/OH-INF/binding/binding.xml]: Thanks for the feedback and the patch. I've updated as requested.

File: bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexServerHandler.java
Comment (lsiepel) [reply bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexServerHandler.java:270]: Like fwolter said; as pollingJob can be changed outside the scope of this method. A race condition might happen that in between those two checks it changes from non-null to null. I agree that this might never ever happen, but it is a best practice to create a local var for this case.

**Comment by jlaur** [comment]
I have added "awaiting feedback" because there is still an unresolved comment for quite some time: https://github.com/openhab/openhab-addons/pull/12135#discussion_r933994350

File: bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexPlayerHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexPlayerHandler.java:68]:
```suggestion
```

File: bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexServerHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.plex/src/main/java/org/openhab/binding/plex/internal/handler/PlexServerHandler.java:62]:
```suggestion
```

**Comment by mlobstein** [comment]
How long does this remain in `awaiting feedback` status before being closed? Although I don't use plex, I wouldn't mind seeing this make it into the official distribution. I would be willing to finish the changes needed to get this done.

**Comment by jlaur** [comment]
> How long does this remain in `awaiting feedback` status before being closed? Although I don't use plex, I wouldn't mind seeing this make it into the official distribution. I would be willing to finish the changes needed to get this done.

@aronbeurskens - would that be okay with you? And @fwolter, do you think you would be able to finish the review if @mlobstein would help resolving the last comments?

**Comment by fwolter** [comment]
Yes!

**Comment by aronbeurskens** [comment]
Of course, no problem. When merged its also easier for me to maintain

**Comment by mlobstein** [comment]
replaced by #15057

**Pull request closed without merging**
